File: adaptive_repair/data/python_programs/detect_cycle.py
def detect_cycle(node):
    # Handle empty list case
    if node is None:
        return False

    hare = tortoise = node

    while True:
        # Before moving, ensure the hare can move two steps without encountering None.
        # If hare.successor is None, the list ends (no cycle).
        # If hare.successor.successor is None, hare is at the second-to-last node,
        # and moving two steps would attempt to access 'successor' of None (the last node's successor),
        # leading to an AttributeError. In this case, the list also ends (no cycle).
        if hare.successor is None or hare.successor.successor is None:
            return False

        tortoise = tortoise.successor
        hare = hare.successor.successor

        if hare is tortoise:
            return True

File: adaptive_repair/data/java_programs/DETECT_CYCLE.java
package java_programs;
import java.util.*;
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

/**
 *
 * @author derricklin
 */
public class DETECT_CYCLE {
    public static boolean detect_cycle(Node node) {
        if (node == null) { // Handle empty list
            return false;
        }

        Node hare = node;
        Node tortoise = node;

        while (true) {
            // Tortoise moves one step
            tortoise = tortoise.getSuccessor();

            // Hare attempts to move two steps
            // Check if hare can move one step
            if (hare.getSuccessor() == null) {
                return false; // End of list reached, no cycle
            }
            // Store the result of the first step to check the second step
            Node hareNext = hare.getSuccessor();
            // Check if hare can move a second step
            if (hareNext.getSuccessor() == null) {
                return false; // End of list reached, no cycle
            }
            hare = hareNext.getSuccessor(); // Hare moves two steps

            if (hare == tortoise) {
                return true; // Cycle detected
            }
        }
    }
}